torch/csrc/jit/codegen/cuda/expr_simplifier.h

```diff
@@ -0,0 +1,533 @@
+//
+// Theorem 6.5: If -|a| < r < |a|, then r % a = r, r / a = 0
+// Proof: This can be proved directly from Definition 0
+//
+// Theorem 6.6: a/(-b) = -a/b, a%(-b) = -a%b
+// Proof: See "Implementations of Div and Mod", this is a written in the paper
+//
+// Theorem 6.7: If compatible_sign(a, b), then (a + b) % c = (a % c + b % c) % c
+// where compatible_sign(a, b) is defined as ab>=0
+// Proof: According to Theorem 6.3, this is just to prove
+// a + b = a % c + b % c (mod c)
+// Because of Theorem 6.4, we have a = a % c (mod c), b = b % c (mod c),
+// applying Theorem 3.3, we get what we want
+//
+// Theorem 6.8: If compatible_sign(a, b), then (a * b) % c = (a % c * b % c) % c
+// where compatible_sign(a, b) is defined as ab>=0
+// Proof: Similar to above
+//
+// Theorem 6.9: If a is a multiple of b, then a % b = 0
+// Proof:  This can be proved directly from Euclid's division lemma
+//
+// Theorem 6.10: If b is a multiple of c, then we have: a*(b/c) = (a*b)/c
+// Proof: Same proof as 4.10
+//
+// Theorem 6.11: a/(b*c) = (a/b)/c
+// Proof: This is part of Theorem 6.1
+//
+// Theorem 6.12: a % (b * c) = a % b + ((a / b) % c) * b
+// Proof: Already proved in the proof of Theorem 6.1
+//
+// Theorem 6.13: If d divides a and b, then a % b = ((a / d) % (b / d)) * d
+// Proof: Same proof as 4.13
+//
+// Theorem 6.14: If b is a multiple of c, then a/(b/c) = (a*c)/b
+// Proof: Same proof as 4.14
+//
+// Theorem 6.15: If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+// (a+b)/c = a/c + b/c
+// Proof: From Theorem 6.7 and Theorem 6.5
+// (a + b) % c = (a % c + b % c) % c = a % c + b % c
+// The rest of the proof is the same as 4.15
+//
+// 7. Summary
+// ==========
+//
+// In summary, in nvFuser, we can use the following rules to simplify integer
+// expressions:
+//
+// A) Associativity of +: a + (b + c) = (a + b) + c
+// B) Associativity of *: a * (b * c) = (a * b) * c
+// C) Commutativity of +: a + b = b + a
+// D) Commutativity of *: a * b = b * a
+// E) Distributivity of * over +: a * (b + c) = (a * b) + (a * c)
+// F) Distributivity of * over +: (a + b) * c = (a * c) + (b * c)
+// G) (-a) / b = -(a / b) = a / (-b)
+// H) (-a) % b = -(a % b) = a % (-b)
+// I) If -|a| < r < |a|, then r % a = r, r / a = 0
+// J) Distributivity of % over +:
+//    If compatible_sign(a, b), then (a + b) % c = (a % c + b % c) % c
+// K) Distributivity of % over *:
+//    If compatible_sign(a, b), then (a * b) % c = (a % c * b % c) % c
+// L) If a is a multiple of b, then a % b = 0
+// M) If b is a multiple of c, then we have: a*(b/c) = (a*b)/c
+// N) a / (b * c) = (a / b) / c
+// O) If d divides a and b, then a % b = ((a / d) % (b / d)) * d
+// P) If b is a multiple of c, then a/(b/c) = (a*c)/b
+// Q) If compatible_sign(a, b) and -|c| < a % c + b % c < |c|, then
+//    (a+b)/c = a/c + b/c
+//
+// Some examples on applying the above rules to simplify expressions:
+//
+// Example 7.1: Given that a >= 0 and b >= 0, simplify (a*4 + b) % 4
```

Comment:
Should we try to do better at using `size_t` instead of `int64_t` in code generation? Or do you think it may be sufficient that we apply these optimizations even if the compiler may not be able to.

Reply:
I would write a prover to prove `a >= 0` and `b >= 0` from the definition of `a` and `b`. To begin with, the prover can be very simple: if x >= 0, y >= 0, then x + y  >= 0, x * y >= 0, x / y >= 0, x % y >= 0. This simple prover should be able to cover our cases, but if later we want more, we can improve the prover.